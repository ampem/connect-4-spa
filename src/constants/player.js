export const RED_PLAYER = 'RED_PLAYER';
export const YELLOW_PLAYER = 'YELLOW_PLAYER';