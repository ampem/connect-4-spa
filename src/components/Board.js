import React, { Component } from 'react';
import keys from 'lodash/keys';
import { RED_PLAYER, YELLOW_PLAYER } from '../constants/player';
import { isDraw, isWinningMove } from '../utils/gameLogic.js';
import Cell from './Cell';
import Column from './Column';
import styles from '../styles/board.css';

class Board extends Component {
    constructor(props) {
        super(props);
        this.state = {
            currentPlayer: RED_PLAYER,
            winner: false,
            draw: false,
            columns: [[],[],[],[],[],[],[]]
        };
    }

    renderColumns() {
        const columns = this.state.columns;
        return keys(columns).sort().map( (_,i) => <Column
                key={i}
                cells={columns[i]}
                clickHandler={() => this.addCellToColumn(i)}
            /> );
    }

    addCellToColumn(columnId) {
        if(this.state.columns[columnId].length > 5) {
            return;
        }
        if(this.state.winner || this.state.draw){
            return;
        }
        const currentPlayer = (this.state.currentPlayer === RED_PLAYER) ? YELLOW_PLAYER : RED_PLAYER;
        const columns = {
            ...this.state.columns,
            [columnId]: [...this.state.columns[columnId], currentPlayer]
        };
        const rowId = this.state.columns[columnId].length;
        const winner = isWinningMove(columnId, rowId, columns);
        const draw = isDraw(columns);
        this.setState({columns, currentPlayer, winner, draw});
    }

    render() {
        const nextPlayer = (this.state.currentPlayer === RED_PLAYER) ? YELLOW_PLAYER : RED_PLAYER;
        const winner = this.state.winner;
        const draw = this.state.draw;

        return (
            <section>
                <div className="board-mask"></div>
                <div className="board-columns">
                    {this.renderColumns()}
                </div>
                { !winner &&
                    <div className="board-player">Next Player:
                        <Cell occupiedBy={nextPlayer} />
                    </div>
                }
                { winner && <div className="board-winner-message">You WON!!!</div> }
                { draw && <div className="board-draw-message">Draw!!!</div> }
            </section>
        );
    }
}

export default Board;
