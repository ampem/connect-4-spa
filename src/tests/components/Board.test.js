import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { RED_PLAYER, YELLOW_PLAYER } from '../../constants/player';
import Board from '../../components/Board';


describe('Board', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<Board />);
        expect(wrapper).toMatchSnapshot();
    });
    it('renders `win` message on win condition', () => {
        const wrapper = shallow(<Board />);
        wrapper.setState({winner: RED_PLAYER});
        expect(wrapper.find('div.board-winner-message').length).toBe(1);
    });
    it('hides next player info on win condition', () => {
        const wrapper = shallow(<Board />);
        wrapper.setState({winner: RED_PLAYER});
        expect(wrapper.find('div.board-player').length).toBe(0);
    });
    it('shows the `draw` message on `draw` condition', () => {
        const wrapper = shallow(<Board />);
        wrapper.setState({
            draw: true
        });
        expect(wrapper.find('div.board-draw-message').length).toBe(1);
    });

    describe('addCellToColumn', () => {
        it('Does not add over 6 markers', () => {
            const wrapper = shallow(<Board />);
            wrapper.setState({
                currentPlayer: RED_PLAYER,
                winner: null,
                columns: [Array(6).fill(RED_PLAYER),[],[],[],[],[],[]]
            });
            wrapper.instance().addCellToColumn(0);
            expect(wrapper.state().columns[0].length).toBe(6)
        });
        it('Does nothing if the game is already won', () => {
            const wrapper = shallow(<Board />);
            wrapper.setState({
                winner: RED_PLAYER
            });
            wrapper.instance().addCellToColumn(0);
            expect(wrapper.state().columns[0].length).toBe(0)
        });
        it('Adds the next marker to the column', () => {
            const wrapper = shallow(<Board />);
            wrapper.setState({
                currentPlayer: RED_PLAYER,
                winner: false,
                draw: false,
                columns: [[RED_PLAYER],[],[],[],[],[],[]]
            });
            wrapper.instance().addCellToColumn(0);
            expect(wrapper.state().columns).toEqual({
                    '0':[RED_PLAYER, YELLOW_PLAYER],
                    '1':[],
                    '2':[],
                    '3':[],
                    '4':[],
                    '5':[],
                    '6':[]
                });
        });
        it('Alternates currentPlayer after adding a marker', () => {
            const wrapper = shallow(<Board />);
            wrapper.setState({
                currentPlayer: RED_PLAYER,
                winner: false,
                columns: [[RED_PLAYER],[],[],[],[],[],[]]
            });
            wrapper.instance().addCellToColumn(0);
            expect(wrapper.state().currentPlayer).toBe(YELLOW_PLAYER);
            wrapper.instance().addCellToColumn(0);
            expect(wrapper.state().currentPlayer).toBe(RED_PLAYER);
        });
    });
});
