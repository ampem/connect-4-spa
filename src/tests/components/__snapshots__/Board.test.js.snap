// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Board />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="board-mask"
/>,
        <div
          className="board-columns"
>
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
          <Column
                    cells={Array []}
                    clickHandler={[Function]}
          />
</div>,
        <div
          className="board-player"
>
          Next Player:
          <Cell
                    occupiedBy="YELLOW_PLAYER"
          />
</div>,
        false,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "board-mask",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
            <Column
              cells={Array []}
              clickHandler={[Function]}
/>,
          ],
          "className": "board-columns",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "6",
            "nodeType": "function",
            "props": Object {
              "cells": Array [],
              "clickHandler": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Next Player:",
            <Cell
              occupiedBy="YELLOW_PLAYER"
/>,
          ],
          "className": "board-player",
        },
        "ref": null,
        "rendered": Array [
          "Next Player:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "occupiedBy": "YELLOW_PLAYER",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      false,
      false,
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="board-mask"
/>,
          <div
            className="board-columns"
>
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
            <Column
                        cells={Array []}
                        clickHandler={[Function]}
            />
</div>,
          <div
            className="board-player"
>
            Next Player:
            <Cell
                        occupiedBy="YELLOW_PLAYER"
            />
</div>,
          false,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "board-mask",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
              <Column
                cells={Array []}
                clickHandler={[Function]}
/>,
            ],
            "className": "board-columns",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "6",
              "nodeType": "function",
              "props": Object {
                "cells": Array [],
                "clickHandler": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Next Player:",
              <Cell
                occupiedBy="YELLOW_PLAYER"
/>,
            ],
            "className": "board-player",
          },
          "ref": null,
          "rendered": Array [
            "Next Player:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "occupiedBy": "YELLOW_PLAYER",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        false,
        false,
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
